File: kt/Minsk/minsk/src/main/kotlin/dev/phytolizer/minsk/analysis/binding/BoundUnaryOperator.kt
package dev.phytolizer.minsk.analysis.binding

import dev.phytolizer.minsk.analysis.syntax.SyntaxKind
import java.lang.reflect.Type

internal class BoundUnaryOperator private constructor(
    private val syntaxKind: SyntaxKind,
    val kind: BoundUnaryOperatorKind,
    private val operandType: Type,
    val resultType: Type,
) {
    companion object {
        private val operators = listOf(
            BoundUnaryOperator(
                SyntaxKind.PlusToken,
                BoundUnaryOperatorKind.Identity,
                Int::class.java,
                Int::class.java
            ),
            BoundUnaryOperator(
                SyntaxKind.MinusToken,
                BoundUnaryOperatorKind.Negation,
                Int::class.java,
                Int::class.java
            ),
        )

        fun bind(syntaxKind: SyntaxKind, operandType: Type): BoundUnaryOperator? {
            return operators.firstOrNull { it.syntaxKind == syntaxKind && it.operandType == operandType }
        }
    }
}

File: nim/src/minsk/codeAnalysis/binding/expressions/boundUnaryOperator.nim
import std/options

import minsk/minskObject
import minsk/codeAnalysis/syntax/syntaxKind
import sequtilsExt/sequtilsExt

type
  BoundUnaryOperatorKind* {.pure.} = enum
    Identity
    Negation
    LogicalNegation

  BoundUnaryOperator* = object
    syntaxKind: SyntaxKind
    kind*: BoundUnaryOperatorKind
    operandType: MinskObjectKind
    resultType*: MinskObjectKind

const
  BOUND_UNARY_OPERATORS = [
    BoundUnaryOperator(
      syntaxKind: SyntaxKind.PlusToken,
      kind: BoundUnaryOperatorKind.Identity,
      operandType: mokInteger,
      resultType: mokInteger,
    ),
    BoundUnaryOperator(
      syntaxKind: SyntaxKind.MinusToken,
      kind: BoundUnaryOperatorKind.Negation,
      operandType: mokInteger,
      resultType: mokInteger,
    ),
    BoundUnaryOperator(
      syntaxKind: SyntaxKind.BangToken,
      kind: BoundUnaryOperatorKind.LogicalNegation,
      operandType: mokBoolean,
      resultType: mokBoolean,
    ),
  ]

proc bindUnaryOperator*(
  kind: SyntaxKind,
  operandType: MinskObjectKind
): Option[BoundUnaryOperator] =
  sequtilsExt.find(
    BOUND_UNARY_OPERATORS,
    proc(x: BoundUnaryOperator): bool =
      x.syntaxKind == kind and x.operandType == operandType
  )
